cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AccessControllerV2.java

```diff
@@ -0,0 +1,127 @@
+import io.cdap.cdap.proto.security.Permission;
+import io.cdap.cdap.proto.security.Principal;
+import io.cdap.cdap.proto.security.Role;
+
+import java.util.Properties;
+import java.util.Set;
+import java.util.jar.Attributes;
+
+/**
+ * Interface for managing {@link Principal principals'} authorization for {@link Permission
+ * permissions} on {@link EntityId CDAP entities}. Authorization extensions must implement this
+ * interface to delegate authorization to appropriate authorization back-ends. The contract with
+ * Authorization extensions is as below:
+ *
+ * <ul>
+ *   <li>Authorization is enabled setting the parameter {@code security.authorization.enabled} to true in
+ *   {@code cdap-site.xml}.</li>
+ *   <li>The path to the extension jar bundled with all its dependencies must be specified by
+ *   {@code security.authorization.extension.jar.path} in cdap-site.xml</li>
+ *   <li>The extension jar must contain a class that implements {@link AccessControllerV2}. This class must be
+ *   specified as the {@link Attributes.Name#MAIN_CLASS} in the extension jar's manifest file.</li>
+ *   <li>The contract with the class that implements {@link AccessControllerV2} is that it must have a default
+ *   constructor.</li>
+ *   <li>{@link AccessControllerV2} also provides lifecycle methods for extensions.
+ *   {@link #initialize(AuthorizationContext)} can be used to perform initialization tasks. This method provides
+ *   an {@link AuthorizationContext} which gives extensions access to CDAP entities for operations like creating
+ *   and accessing datasets, accessing datasets in transactions, etc. It also provides access to
+ *   {@link Properties extension properties} via the {@link AuthorizationContext#getExtensionProperties()} method.
+ *   The {@link Properties} object returned form this method is populated with all configuration settings from
+ *   {@code cdap-site.xml} that have keys with the prefix {@code security.authorization.extension.config}.</li>
+ *   <li>The {@link #destroy()} method can be used to perform cleanup tasks.</li>
+ * </ul>
+ *
+ * This is newer version of {@link AccessController}
+ */
+@Beta
+public interface AccessControllerV2 extends GrantFetcher, PermissionManager, AccessEnforcerV2 {
+
+  /**
+   * Initialize the {@link AccessControllerV2}. Authorization extensions can use this method to access
+   * an {@link AuthorizationContext} that allows them to interact with CDAP for operations such as
+   * creating and accessing datasets, executing dataset operations in transactions, etc.
+   *
+   * @param context the {@link AuthorizationContext} that can be used to interact with CDAP
+   */
+  default void initialize(AuthorizationContext context) {
+    // default no-op implementation
+  }
+
+  /**
+   * Create a role.
+   *
+   * @param role the {@link Role} to create
+   * @throws AlreadyExistsException if the the role to be created already exists
+   */
+  void createRole(Role role) throws AccessException;
```

Comment:
If we continue supporting this interface, we must do it in CAL-friendly way as well with EnforcementResult. And similarly other calls. I would probably create 2 separate interfaces for this for API/SPI instead of having it in the AccessController. And we'll need proper CAL wrapper for SPI->API